Add tensor parallel sharding APIs for torch export

feikou opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111236

Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

feikou committed:
Add tensor parallel sharding APIs

Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

[ghstack-poisoned]

feikou commented:
/easycla

feikou committed:
Update on "Add tensor parallel sharding APIs"


Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

[ghstack-poisoned]

feikou commented:
/easycla

feikou committed:
Update on "Add tensor parallel sharding APIs"


Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/experimental/tensor_parallel.py:
```diff
@@ -0,0 +1,432 @@
+    return tuple(input_specs_list)
+
+
+def _get_arg_schema(
```
nit: maybe call this `_get_op_schema` and directly construct OpSchema instead

wanchaol reviewed (commented):
This looks pretty good! I left a few nits, the main thing we should figure out is that we probably should not assume the even/odd layer of Linear and apply col/row parallel there, let's try to see if we can allow some level of module annotation in the exported program for better UX.

Another possible approach that I recently discussed with @SherlockNoMad  is that we can possibly first replace the input arguments with DTensor after the module annotation, then "re-export" the graph by going through the DTensor's torch_dispatch logic. But I feel we can make that a follow up change.

feikou commented on test/distributed/_tensor/experimental/test_tensor_parallel.py:
```diff
@@ -0,0 +1,61 @@
+            )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = torch.chunk(x, 2, dim=1)[0]
```
This was to test special handling of getitem op in mark_sharding.

feikou commented on test/distributed/_tensor/experimental/test_tensor_parallel.py:
```diff
@@ -0,0 +1,61 @@
+        torch.manual_seed(0)
+        model = MLPListModule(2).to(device=self.device_type)
+        inputs = (torch.randn((10, 12)).to(device=self.device_type),)
+        with torch.inference_mode():
```
Currently this is pretty commonly used in model processing and inference since gradient calculation is not needed. I'm curious why some aten ops may not decompose under inference mode. Is this something WIP or by design?

wanchaol replied:
Something it's a bit weird under torch.inference_mode(), for example, if you capture ops inside inference_mode, it would give you either `aten.linear.default` or `aten.matmul.default`, instead of giving you `aten.mm/addmm.default` instead. That is probably ok if torch.export is not run under `inference_mode`

feikou replied:
Interesting, that's good to know.

feikou commented on torch/distributed/_tensor/experimental/tensor_parallel.py:
```diff
@@ -0,0 +1,432 @@
+                    placements=(Replicate(),),
+                )
+    # If there are odd number of linear layers, no need to shard the params of the last layer.
+    if len(linear_node_names) % 2 == 1:
```
Sounds good. Will follow up to make tagging more generic.

feikou commented on torch/distributed/_tensor/experimental/tensor_parallel.py:
```diff
@@ -0,0 +1,432 @@
+                output_sharding = DTensor._propagator.propagate_op_sharding(
+                    op_schema,
+                )
+                # The stride from output_sharding may be incorrect for addmm (https://fburl.com/code/kb1ko758),
```
Yes, tried that the PR worked well for this case!

feikou committed:
Update on "Add tensor parallel sharding APIs"


Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/experimental/tensor_parallel.py:
```diff
@@ -0,0 +1,426 @@
+    exported_program: ExportedProgram,
+    rank: int,
+    world_size: int,
+) -> ExportedProgram:
```
Let's add a docstr to:
* explicitly mention that this is a graph transformation pass for torch.export to transform a single device graph to a tensor parallel graph
* add a warning here that this API is experimental and subject to change so that people don't abuse this API

wanchaol commented on torch/distributed/_tensor/experimental/tensor_parallel.py:
```diff
@@ -0,0 +1,426 @@
+import copy
```
nit: let's also rename this file to sth like `tp_transform.py` or `tensor_parallel_transform.py` to make this more explicit that this is a graph transformation pass.

wanchaol commented on torch/distributed/_tensor/experimental/tensor_parallel.py:
```diff
@@ -0,0 +1,426 @@
+        return PassResult(gm, True)
+
+
+def mark_tensor_parallel_shardings(
```
maybe we should made all those APIs be private (i.e. by prepanding `_`) as they are all exclusive to this file only. let's only leave a public `tensor_parallel_transformation` API and make the rest to be private

wanchaol reviewed (approved):
lgtm, great work! Left a few comments to add some docstr and make some APIs be private.

feikou committed:
Update on "Add tensor parallel sharding APIs for torch export"


Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

[ghstack-poisoned]

feikou committed:
Update on "Add tensor parallel sharding APIs for torch export"


Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

[ghstack-poisoned]
